### ISSUE TITLE
trial runs tests from .pyc files even if there is no .py file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2036]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2036|
|Type|defect|
|Created|2006-08-30 03:28:41Z|
|Branch|https://github.com/twisted/twisted/tree/trial-stale-pyc-2036-2|

This is generally annoying.  Often .pyc files are left around after the .py has been intentionally deleted.  Then, test runs fail because the .pyc file tries to do something that mostly likely doesn't work anymore.

Attachments:

* [2036.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ad7/ad707ee02cf3c3a30babb2e4259100a414e7ca67/a31ff94e5ab2fd996be8121872cebca16ca49293.patch) (479 bytes) - added by michaelnt on 2011-01-31 03:07:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2036 2036
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_trial_stale_pyc_2036_2 branches/trial-stale-pyc-2036-2
branch_author__washort__glyph washort, glyph
status__closed closed
resolution__wontfix wontfix
component__trial trial
keywords__ 
time__1156908521000000 1156908521000000
changetime__1504487302445402 1504487302445402
version__None None
owner__ 
cc__therve cc__thijs cc__jesstess cc__michaelnt cc__davidsarah
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

To clarify, I don't really care if trial lets you do "trial foo.bar.baz" when baz.py has been deleted but baz.pyc still exists.  What I really don't want is for "trial foo" to run the tests in baz.pyc if there is no baz.py.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

What if there are only .pycs, intentionally?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

I think that's a pretty bad place to be.  Don't do that?  There could be a --run-sourceless-tests flag, though, I suppose.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Any chance you can make this a priority?  I can prioritize a ticket or two which you are interested in seeing completed in exchange.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

Sure.

I can't think of anything I really want prioritized that you are likely to do before Sol goes nova. Fixing broken tests always makes me happy, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

In our last IRC conversation, we concluded this is the feature which should be added:

 * If trial finds a ```.pyc``` without a ```.py``` beneath a directory it was told to search for tests, it will refuse to run and report what it found.
 * There should be a trial option to revert to the old behavior and ignore this error if that is really what you mean.

In the future - as a separate step - it may also be desirable to include some tool, perhaps also a trial option, to "clean up" working copies and remove pycs to force recompilation.  This might be more appropriate for a tool like Combinator, however.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_7"></a>|@washort commented|
|-|-|

(In [26110]) Branching to 'trial-stale-pyc-2036'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

how about .pyo?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

and uppercase extensions.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_10"></a>|@washort commented|
|-|-|

trial doesn't load .pyo files if no .py or .pyc exists for them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Even if trial is run with Python with -O?

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_12"></a>|@washort commented|
|-|-|

Even then. Also, python doesn't import files with uppercase extensions.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="trickos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|trickos commented|
|-|-|

Replying to [washort](#note_13):

Here is what I read from the fix:

- Abort if .pyc with no matching .py

- Add an option to ignore the former behaviour

Am I right ? I am not sure this new behaviour is less annoying than the current one: right now we have random failures, with the patch we have non-random failures. But in both case it's up to the user to cleanup the test directory.

Wouldn't it be possible to ignore the tests with no matching .py ?

For the record, I hit this while running a bisection on a twisted base project. It means that at every bisection iteration I have to cleanup the .pyc. This is very easy to reproduce as you can see and very annoying.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Unless it's possible to tell Python itself to ignore .pyc files, I don't think trial can ignore them but still run the test suite.  The problem comes when you have an indirect usage of a pyc file.  Say test_foo.py imports from test_bar - if you have test_bar.pyc but no test_bar.py, trial can ignore the test_bar module, but when it runs test_foo.py, test_bar gets imported anyway.  This is still probably not very desirable behavior since it still doesn't reflect what you'd get if you had a "clean" version of the source.

I had originally just wanted things to be ignored by trial, but I think because of this behavior that's not reasonable.

**Comment by twisted-trac** [comment]
|<img alt="trickos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|trickos commented|
|-|-|

Fair enough. Perhaps hacking into the importer may help but it sound like a complicated option.

I have additional remarks about the propose changed:

- It would be great to tell what .pyc is causing the run to abort (just list it in the exception)

- It seems trial also aborts even if the stale module is not a test module. This was unexpected even if I can imagine reasons (though a bit paranoid ones) to do that.

In comparison, nosetests does not suffer from this issue (and yes trial output is much nicer, nosetests does not handle skip exceptions... but still).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

Is nosetest better in the ways you describe (tells you which .pyc caused the abort, doesn't abort on stale non-test files) or in other ways too?  If other ways, can you give a quick summary?

**Comment by twisted-trac** [comment]
|<img alt="trickos's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|trickos commented|
|-|-|

I don't pretend I know nosetests very well, I just use it on another project. I think the difference is it discovers tests by walking the file system instead of using the import machinery, so it does not see .pyc ghosts. But it is vulnerable to the kind of broken imports you described.

I don't know a lot of tests system, but one which does not suffer from this is Mercurial one which installs a temporary copy of itself in a temp directory before running the tests. This ensures a very good isolation of the test environment but is more expensive to run.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

OK the branch looks good indeed, but I agree with trickos that the behavior is a bit boring. The Zope test runner deletes the pyc files, and I think this a good default. Unfortunately it may break some stuff. My suggestion:

 * add a `--remove-stale-pyc` option, which deletes pyc files when enabled.
 * add a `--keep-stale-pyc` option, which is a no-op now.
 * when the remove option is not enabled, that the keep options is not passed, and that a pyc file is detected, print a deprecation warning: "Pyc file detected. In twisted XX this file will be automatically detected unless you pass --keep-stale-pyc".

What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set owner to @washort|
|-|-|

  1. Still doesn't support .pyos, as itamar mentioned earlier.
  1. YOU ALREADY HAVE A FILEPATH WHY ARE YOU USING `os.path`!  DO NOT DO THAT
    1. ZOMG DO NOT ''EVER'' USER SPLITEXT!!!
  1. You should really be modifying `twisted.python.modules` to provide some abstraction over these operations; like `.isCompiled()` or `.compiledPath()` or something.  Maybe even `.deleteCompiled()` or something.

I generally like therve's suggestion for the user interface.  However, there is a corner case which I think we should also support: sometimes _all you have_ are .pyc files.  For example, if you've compiled a crushed-down version of your application into a .zip file for deployment on a mobile device, and you want to run the tests in the deployment.  Or, if you're testing a (bleh) proprietary application that doesn't ship source.  I think if there are no `.py` files anywhere (for a suitable definition of "anywhere") we should not attempt to clean up the .pycs.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_21"></a>|@washort removed owner|
|-|-|

Since test loading from a package is dependent upon importing the package first, I'm not at all clear how to directly test loading a module from a .pyo file without requiring "python -O" be used.

 * Got rid of splitext and os.path stuff.
 * Moved the path checks and such into `twisted.python.modules.PythonModule.hasSource()`.
 * Added `--keep-stale-pyc` and `--remove-stale-pyc` options.
 * `ZipPath` doesn't support deletion, so pyc removal isn't done for zip packages.
 * I think the deprecation and defining of "anywhere" should be done in separate tickets.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph set owner to @washort<br>@glyph set status to `new`|
|-|-|

  1. `hasSource` should have more tests.
    1. the test case should be broken apart into pieces so that it tests individual features one at a time.  `test_hasSource` is mixing together setup and verification in a big multi-step mess.
    1. the test case should be written in such a way that it tests both `ZipPath`-backed and `FilePath`-backed path entries.  `PathModificationTest` may be used, or may be useful as an example.  Right now the implementation is wrong, because it statically creates a `FilePath` from a path string retrieved from another path.  (!?!?!?!!?!!! why do you keep doing that!!)
  1. `hasSource()` should really be `sourcePath()`; `hasSource()` could then be trivially implemented as `sourcePath().exists()`, whereas the inverse is not possible.
  1. The new flags being passed down through `loadByNames`, etc, are problematic:
    1. they are undocumented at various levels.
    1. they violate the existing convention, which I think is better than passing a big stack of flags around.  For other options, `_getLoader` pulls out the relevant options from the configuration and supplies them as ... erm ... argments to `TestLoader`, sort of.  Ideally TestLoader would actually accept these as constructor arguments rather than being randomly mutated.  The loader should hold all the state responsible for loading as attributes, rather than as an interaction with the 'script' module, on the stack.  (Sorry about 'recurse'; that set a bad precedent.  I hope it gets rehabilitated eventually as well.)
    1. they are only honored by loadPackage rather than all the other ways you might load some files.  If you specify something inside a module, you don't get any warning if the test module is now a stale .pyc, but if you specify the package the module is in you will get a warning.
  1. loadPackage is starting to get pretty complex.  As the discovery is growing more and more features, it should really be put into a separate method from the suite-building stuff.  I suggest putting the stuff that builds the 'discovered' list into a nested list of generators.
  1. `isinstance(disco.filePath, FilePath)` is a crummy way to test for the presence of 'remove'; I would recommend a `getattr()` check instead, so that this stuff can be tested and run with in-memory filepath implementations.
  1. remove() might raise an exception if, for example, you don't have write access to the path.  I think that should be handled.  You could address both this point and the previous point by implementing remove() on `ZipPath` and making it behave the same as a `FilePath.remove` that you don't have permission to access.
  1. forceStale's documentation is really confusing, and its name strikes me as backwards.  hopefully you'll be moving it to an attribute, but when you do so, please change its documentation to use fewer words like 'not' and 'no'.
  1. really minor stuff:
    1. 'pyc' is overused, since the files in question could be pycs or pyos or (hopefully, one day) .class files or CLR assemblies.  I think 'bytecode' or 'compiled' might be a better word.
    1. 'stale' pycs, in my mind, are those which are out-of-date from the .py files they correspond to.  'orphaned' pycs, maybe?
  1. I also feel like there's an implicit problem here, where the default behavior is going to cause inscrutable errors if you're actually trying to test packaged .pyc files (for example, packaged for testing in some embedded environment or self-updating desktop application).  I would like it if package discovery would note whether _any_ source files exist for it (or, perhaps as an approximation, whether the package *itself* has a source file, i.e. is `__init__.py` present) before warning about 'stale' (or 'orphaned') pycs.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_24"></a>|@washort removed owner|
|-|-|

Quite a list.

I believe I've addressed all of this. Specifically:

 1. `twisted.test.test_modules.PathModificationTests` now has `test_sourcePath` and `test_importedSourcePath`.
 2. `hasSource` is gone, `sourcePath` added.
 3. Flag arguments removed from all those methods, `forceOrphan` and `removeOrphan` attributes on TestLoader now.
 4. I moved _some_ stuff to a different method, `_loadModulesNoOrphans`. I am not entirely pleased with how that went. Part of the problem is that these loader methods all need to return a test suite, so raising an exception in the loader method when encountering an error isn't feasible. Also, since it might need to remove a .pyc file, the caller has to examine its output. Further improvements might require extensive reorganization.
 5. check done with `getattr` now.
 6. Exceptions raised by `remove` are discarded. I couldn't think of a reasonable place to report them, and no error is reported for paths with no `remove` method, so this seemed reasonable.
 7. Documentation reworded.
 8. Only the test methods refer to '.pyc files' now. Variables and docs refer to 'orphan' rather than 'stale' modules.
 9. If `__init__.py` is not present when loading a test package, `removeOrphan` is ignored and `forceOrphan` is enabled. It just now strikes me that maybe a similar check should be when loading individual modules, not packages, but it's not clear what should be checked in that case.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jerub<br>@jerub set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_26"></a>|@jerub commented|
|-|-|

I have reviewed this branch, run the tests, checked coverage, and am mostly happy with it.

I dislike the ```try: remove() except: pass``` - bare excepts are terrible. But I can't think of a better way of handling this either. Perhaps we should log the exception?

Please address this concern - either by making a change to add a logging of the traceback - or tell me why we should just throw away the exception.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set owner to @washort<br>@exarkun set status to `new`|
|-|-|

I don't see why this should still be assigned to Jerub.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_29"></a>|@jesstess commented|
|-|-|

This is so close to being done so I'd like to push it through. Jerub: your issue was with having a bare exception; it looks like FilePath.remove() can only result in an OSError, so can we just change the block to an ```except OSError: pass```?

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_30"></a>|michaelnt removed owner|
|-|-|

Changes the bare exception to only trap OSError as suggested. This is the error already being raised in the test.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_32"></a>|@glyph commented|
|-|-|

(In [30718]) Branching to 'trial-stale-pyc-2036-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_33"></a>|@glyph commented|
|-|-|

Merged forward, applied patch, [submitted build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/trial-stale-pyc-2036-2).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_34"></a>|@glyph set owner to @washort<br>@glyph set status to `new`|
|-|-|

Thanks for keeping this going, michaelnt.  I'm reassigning to washort because he's written most of the code so far, but if you'd like to respond to the rest of this feedback before he gets to it that would be great.

  1. `sourcePath` is wrong for extension modules.  It returns the path to the `.so` file.  This is bad because client code which is inspecting arbitrary modules might end up trying to load a giant pile of x86 code as Python source.  I would prefer a fake source file, like a `.c` file in the same directory.  If you go that route, explicitly note its semantics in the docstring.
  2. `sourcePath`'s returns `None` when there's no source rather than the non-existent `.py` path.  This makes it annoying to re-scan the directory to see if some source has appeared.  (An odd use case, I know, but most use-cases for this stuff are odd.)  As I suggested before, I still think `sourcePath().exists()` would be a more convenient idiom.
  3. `sourcePath`'s docstring doesn't explain its different return values, or have an explicit `@rtype`.  It might not be obvious to a naive reader that it doesn't return a string.
  4. What about running tests on a directory with orphans where I might not have write permission?  We have `--force-orphan`, `--remove-orphan`, and `--keep-orphan` but no `--skip-orphan`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_36"></a>|@itamarst commented|
|-|-|

Re #4, having an extra option seems unnecessary. If you did --remove-orphans and there's no write permission, just report the error.

I'm confused about --remove-orphan and --keep-orphan, or rather the latter; as a user I would assume the default is to not remove orphans, so what is the point of having --keep-orphan? Or if the default is to remove (which I would dislike, and is not documented), what is the purpose of --remove-orphan?

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_38"></a>|@bitglue commented|
|-|-|

Replying to [glyph](#note_6):
> In the future - as a separate step - it may also be desirable to include some tool, perhaps also a trial option, to "clean up" working copies and remove pycs to force recompilation.  This might be more appropriate for a tool like Combinator, however.

You mean like ```find . -name '*.py[oc]' | xargs rm```, or maybe ```find . -name '*.py[oc]' -exec rm '{}' \;```?

I must say, looking at this ticket for the first time now, I think this feature is frivolous or perhaps harmful. I agree having an orphaned ```.pyc``` around can be annoying, but it's not just tests where it causes problems. Python developers should be familiar with this behavior, and having trial ignore ```.pyc``` while the rest of python doesn't violates the principle of least surprise.

Anyhow as pointed out in the discussion, there are legitimate use cases one might only have a ```.pyc```. The solution to the ```.pyc``` "problem" isn't immediately obvious; maybe you want to delete them, maybe ignore them, maybe ignore and report, or some combination of the above for each module in your project. "In the face of ambiguity, refuse the temptation to guess."

Besides, there's an easy fix for projects that are annoyed by ```.pyc``` files and want them deleted, in the form of the bash one-liners above. Throw it in a Makefile or a shell script that also runs trial and bam, you have your desired behavior. If you really want, you could have a test that asserts there are no orphaned .pyc in your project directory, or write a separate utility to implement your desired behavior, without adding complexity and surprising functionality to trial.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_39"></a>|@glyph commented|
|-|-|

Replying to [indigo](#note_38):
> Replying to [glyph](#note_6):
> > In the future - as a separate step - it may also be desirable to include some tool, perhaps also a trial option, to "clean up" working copies and remove pycs to force recompilation.  This might be more appropriate for a tool like Combinator, however.
> 
> You mean like ```find . -name '*.py[oc]' | xargs rm```, or maybe ```find . -name '*.py[oc]' -exec rm '{}' \;```?
> 
> I must say, looking at this ticket for the first time now, I think this feature is frivolous or perhaps harmful. I agree having an orphaned ```.pyc``` around can be annoying, but it's not just tests where it causes problems. Python developers should be familiar with this behavior, and having trial ignore ```.pyc``` while the rest of python doesn't violates the principle of least surprise.

The more I think about this ticket the more I'm inclined to agree.  Bytecode files are part of python, like it or not.

Still though: Python won't scan for arbitrary pyc files, trial does that on its own.  And Python has '-v', to tell you what's being imported and why, but trial doesn't have any debugging facilities for test discovery; so clearly something's missing here.

Also, many people are totally surprised by this behavior, so "principle of least surprise" isn't a good way to describe this.  "principle of actually pretty consistent once you think about it for a couple of weeks", maybe.

> Anyhow as pointed out in the discussion, there are legitimate use cases one might only have a ```.pyc```. The solution to the ```.pyc``` "problem" isn't immediately obvious; maybe you want to delete them, maybe ignore them, maybe ignore and report, or some combination of the above for each module in your project. "In the face of ambiguity, refuse the temptation to guess."

To be fair, nobody's pointed out a legit use-case for having a _mixture_ of .py and .pyc files.

> Besides, there's an easy fix for projects that are annoyed by ```.pyc``` files and want them deleted, in the form of the bash one-liners above. Throw it in a Makefile or a shell script that also runs trial and bam, you have your desired behavior. If you really want, you could have a test that asserts there are no orphaned .pyc in your project directory, or write a separate utility to implement your desired behavior, without adding complexity and surprising functionality to trial.

I don't think it's unreasonable for trial to include this kind of cleanup utility itself.  The bash one-liners are, first of all, un-portable to other platforms, and second, [nobody should ever write a shell script for any reason ever](http://www.dwheeler.com/essays/fixing-unix-linux-filenames.html).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_40"></a>|@hawkowl set status to `closed`|
|-|-|

The root cause is a Python misfeature (pyc files alongside py files) and is fixed in Python 3.2. (https://www.python.org/dev/peps/pep-3147/#case-3-pycache-foo-magic-pyc-with-no-source). For Python 2.7, mitigations including using tox to install your package before running the tests or setting ```PYTHONDONTWRITEBYTECODE=1``` fix this without altering Trial. Plus, I agree with the assertion that it would make Trial act differently than all other Python software on Python 2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

I'm sad to see this closed as wontfix but don't have the energy to re-state the valid points already made on this ticket.
